### PR TITLE
Add Configure > Reconfigure for App Engine Java 8 runtime

### LINKED ISSUE #2914: Add project Configure > Convert to App Engine Java 8
Simplify the process of moving from App Engine standard for Java 7 to Java 8.

screen shot 2018-03-08 at 11 21 16 am

### PR DESCRIPTION
**Pull request by briandealwis**
Add new _Configure > Reconfigure for App Engine Java 8 runtime_ for App Engine standard Java 7 projects.  Adds a suite of context-menu verification tests, as well as an integration test that the conversion works.

I initially used "Convert to" or "Configure for" but it didn't read well IMHO.

Fixes #2914

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/ProjectContextMenuTest.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+/** Ensure configured menu options are shown for various types of project. */
+public class ProjectContextMenuTest extends BaseProjectTest {
+  @Rule public TestProjectCreator projectCreator = new TestProjectCreator();
+
+  @Test
+  public void testPlainJavaProject() {
+    IProject project = projectCreator.withFacets(JavaFacet.VERSION_1_8).getProject();
+    SWTBotTreeItem selected = SwtBotProjectActions.selectProject(bot, project.getName());
+    assertContextMenuPathNotFound(selected, "Debug As", "2 App Engine");
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/ProjectContextMenuTest.java]: I remember I encountered a problem with the menu number ("2" here) before. It wasn't always the same number depending on environments, so took a different approach in the past. See #2939 for example.

File: plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/StandardConversionsIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import org.eclipse.core.resources.IProject;
+import org.eclipse.jst.common.project.facet.core.JavaFacet;
+import org.eclipse.jst.j2ee.web.project.facet.WebFacetUtils;
+import org.eclipse.swtbot.swt.finder.widgets.SWTBotTreeItem;
+import org.eclipse.wst.common.project.facet.core.IFacetedProject;
+import org.junit.Rule;
+import org.junit.Test;
+
+/** Tests for conversion of various App Engine Standard project types. */
+public class StandardConversionsIntegrationTest extends BaseProjectTest {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/StandardConversionsIntegrationTest.java]: ditto. Won't even need a field and instantiate a `bot` locally.

File: plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/ProjectContextMenuTest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+import org.eclipse.jst.common.project.facet.core.JavaFacet;
+import org.eclipse.jst.j2ee.web.project.facet.WebFacetUtils;
+import org.eclipse.swtbot.swt.finder.exceptions.WidgetNotFoundException;
+import org.eclipse.swtbot.swt.finder.widgets.SWTBotMenu;
+import org.eclipse.swtbot.swt.finder.widgets.SWTBotTreeItem;
+import org.junit.Rule;
+import org.junit.Test;
+
+/** Ensure configured menu options are shown for various types of project. */
+public class ProjectContextMenuTest extends BaseProjectTest {
```
[the remark this answers is not in the document]
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/ProjectContextMenuTest.java]: BaseProjectTest also closes the Welcome view.

File: plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/ProjectContextMenuTest.java (hunk 1 above)
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/ProjectContextMenuTest.java]: Good idea. I started going down that route, but it took a turn towards a cliff and I steered back.  I'll take a look at #2939.

File: plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/src/com/google/cloud/tools/eclipse/appengine/standard/java8/ConvertToJava8RuntimeHandler.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+public class ConvertToJava8RuntimeHandler extends AbstractHandler {
+
+  private static final Path APPENGINE_DESCRIPTOR = new Path("appengine-web.xml"); //$NON-NLS-1$
+
+  @Override
+  public Object execute(ExecutionEvent event) throws ExecutionException {
+    List<IProject> projects = ProjectFromSelectionHelper.getProjects(event);
+    Preconditions.checkArgument(!projects.isEmpty());
+    Job updateJob = new WorkspaceJob(Messages.getString("reconfiguringToJava8")) { // $NON-NLS-1$
+          @Override
```
[the remark this answers is not in the document]
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/src/com/google/cloud/tools/eclipse/appengine/standard/java8/ConvertToJava8RuntimeHandler.java]: That's what I get with google-java-format!

File: plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/ProjectContextMenuTest.java (hunk 1 above)
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/ProjectContextMenuTest.java]: Yeah, this code is ugly and can be improved even if by writing a custom matcher.

**Comment by chanseokoh** [comment]
Oh! Just realized we should disable the menu for a Maven project. This doesn't work, and facets won't change. (We have not been supporting conversion between Java7 and Java 8 anyway.)

File: plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/src/com/google/cloud/tools/eclipse/appengine/standard/java8/messages.properties [hunk 5]
```diff
@@ -0,0 +1,2 @@
+reconfiguringProject=Reconfiguring {0} 
+reconfiguringToJava8=Reconfiguring projects to use App Engine Java 8 runtime
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/src/com/google/cloud/tools/eclipse/appengine/standard/java8/messages.properties]: project or projects?

Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/src/com/google/cloud/tools/eclipse/appengine/standard/java8/messages.properties]: Plural: handy for those with multiple projects.

File: plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/ProjectContextMenuTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/ProjectContextMenuTest.java]: > BaseProjectTest also closes the Welcome view.

…which is handy when running the tests individually.

File: plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    description.appendText("has menu item that ").appendDescriptionOf(menuLabelMatcher);
+  }
+
+  @Override
+  protected boolean doMatch(Object item) {
+    if (item instanceof SWTBotMenu) {
+      try {
+        List<String> menuTexts = ((SWTBotMenu) item).menuItems();
+        return menuTexts.stream().anyMatch(text -> menuLabelMatcher.matches(text));
+      } catch (WidgetNotFoundException ex) {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java]: No one seems to throw this. No error in my IDE.

File: plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        return false;
+      }
+    } else if (item instanceof SWTBotRootMenu) {
+      try {
+        List<String> menuTexts = ((SWTBotRootMenu) item).menuItems();
+        return menuTexts.stream().anyMatch(text -> menuLabelMatcher.matches(text));
+      } catch (WidgetNotFoundException ex) {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java]: ditto.

File: plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        return false;
+      }
+    } else if (item instanceof Menu) {
+      return UIThreadRunnable.syncExec(
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java]: Very handy, as there is no need to pass around `Display`. I should have known this already. I think I can simplify other code, such as `LoginServiceUi`, which accepts a `Display` only to be able to do `Display.asyncExec()`.

File: plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+          () ->
+              Stream.of(((Menu) item).getItems())
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java]: I'm always confused if `Stream.of()` and `Arrays.stream()` will work differently: https://stackoverflow.com/questions/39874242/stream-vs-stream-of

Can you confirm it works as intended?

File: plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java (hunk 6 above)
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java]: Oh good point — I previously was using `findMenu()` and trying to pass in the matcher.

File: plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java (hunk 8 above)
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java]: It's an SWTBot helper, so I don't think it's appropriate for LoginServiceUi unfortunately.

File: plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java (hunk 9 above)
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/MenuMatcher.java]: `Menu#getItems()` returns an array of `MenuItem`s, so it's ok here.  I did use it in `RunAppEngineShortcutTest` but it was superfluous with your rewrite to use `menuItems()`.

**Pull request merged**